x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java

```diff
@@ -150,16 +159,42 @@ protected ChannelHandler getClientChannelInitializer(DiscoveryNode node, Connect
     }
 
     @Override
-    protected InboundPipeline getInboundPipeline(boolean isRemoteClusterServerChannel) {
-        return new InboundPipeline(
-            getStatsTracker(),
-            threadPool::relativeTimeInMillis,
-            isRemoteClusterServerChannel
-                ? new InboundDecoder(recycler, RemoteClusterPortSettings.MAX_REQUEST_HEADER_SIZE.get(settings))
-                : new InboundDecoder(recycler),
-            new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
-            this::inboundMessage
-        );
+    protected InboundPipeline getInboundPipeline(Channel channel, boolean isRemoteClusterServerChannel) {
+        if (false == isRemoteClusterServerChannel) {
+            return super.getInboundPipeline(channel, false);
+        } else {
+            return new InboundPipeline(
+                getStatsTracker(),
+                threadPool::relativeTimeInMillis,
+                new InboundDecoder(recycler, RemoteClusterPortSettings.MAX_REQUEST_HEADER_SIZE.get(settings), SERVER),
+                new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
+                this::inboundMessage
+            ) {
+                @Override
+                protected void headerReceived(Header header) {
+                    if (header.isHandshake() == false) {
+                        // eagerly (before buffering the full request) authenticate all request headers for this type of channel
+                        assert header.isRequest();
+                        // authn is mostly async, avoid buffering anymore data while authn is in progress
+                        channel.config().setAutoRead(false);
+                        // this prevents thread-context changes to propagate beyond the validation, as netty worker threads are reused
+                        try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().newStoredContext()) {
+                            crossClusterAccessAuthenticationService.tryAuthenticate(
+                                header.getRequestHeaders(),
+                                ActionListener.runAfter(ActionListener.wrap(aVoid -> {
+                                    // authn is successful -> NOOP (the complete request will be subsequently authn & authz & audited)
+                                },
+                                    e -> channel.eventLoop()
+                                        .submit(() -> channel.pipeline().fireExceptionCaught(new HeaderValidationException(header, e)))
+                                ), () -> channel.config().setAutoRead(true))
```

Comment:
Let's log the error message in the error branch because some underlying failures are rather quiet, e.g. no credentials found. Just e.getMessage() not the entire stacktrace. Maybe at debug level so that there is always a way to have this failure identified on the server side.

I think `fireExceptionCaught` can possible be called on its own without `submit` since it seems to do that behind the scene if it is not on the eventloop thread. But I am also OK with using submit since it's is likely more canonical.

Also I am not sure what it entails by calling `setAutoRead(true)` once the exception is fired and handled. But it's likely benign. So should be OK.

Reply:
> Let's log the error message in the error branch because some underlying failures are rather quiet, e.g. no credentials found. Just e.getMessage() not the entire stacktrace. Maybe at debug level so that there is always a way to have this failure identified on the server side.

Good point. I've pushed: https://github.com/elastic/elasticsearch/pull/98814/commits/ef131dfb0ad969e528fd1619b0a334d4f474f752